fedopt_guide/cifar10/federated_trainer.py

```diff
@@ -0,0 +1,183 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Runs federated training on various tasks using a generalized form of FedAvg.
+Specifically, we create (according to flags) an iterative processes that allows
+for client and server learning rate schedules, as well as various client and
+server optimization methods. For more details on the learning rate scheduling
+and optimization methods, see `shared/optimizer_utils.py`.
+Uses the tff.learning.build_federated_averaging_process to perform federated
+averaging on the clients.
+"""
+
+import collections
+from typing import Any, Callable, Optional
+
+from absl import app
+from absl import flags
+import tensorflow as tf
+import tensorflow_federated as tff
+
+from fedopt_guide.cifar10 import federated_cifar10
+from optimization.shared import optimizer_utils
+from utils import utils_impl
+
+_SUPPORTED_TASKS = [
+    'cifar10'
+]
+
+with utils_impl.record_hparam_flags() as optimizer_flags:
+  # Defining optimizer flags
+  optimizer_utils.define_optimizer_flags('client')
+  optimizer_utils.define_optimizer_flags('server')
+  optimizer_utils.define_lr_schedule_flags('client')
+  optimizer_utils.define_lr_schedule_flags('server')
+
+with utils_impl.record_hparam_flags() as shared_flags:
+  # Federated training hyperparameters
+  flags.DEFINE_integer('client_epochs_per_round', 1,
+                       'Number of epochs in the client to take per round.')
+  flags.DEFINE_integer('client_batch_size', 20, 'Batch size on the clients.')
+  flags.DEFINE_integer('clients_per_round', 10,
+                       'How many clients to sample per round.')
+  flags.DEFINE_integer('client_datasets_random_seed', 1,
+                       'Random seed for client sampling.')
+
+  # Training loop configuration
+  flags.DEFINE_string(
+      'experiment_name', None, 'The name of this experiment. Will be append to '
+      '--root_output_dir to separate experiment results.')
+  flags.mark_flag_as_required('experiment_name')
+  flags.DEFINE_string('root_output_dir', '/tmp/fed_opt/',
+                      'Root directory for writing experiment output.')
+  flags.DEFINE_integer('total_rounds', 200, 'Number of total training rounds.')
+  flags.DEFINE_integer(
+      'rounds_per_eval', 1,
+      'How often to evaluate the global model on the validation dataset.')
+  flags.DEFINE_integer('rounds_per_checkpoint', 50,
+                       'How often to checkpoint the global model.')
+  flags.DEFINE_integer(
+      'rounds_per_profile', 0,
+      '(Experimental) How often to run the experimental TF profiler, if >0.')
+
+with utils_impl.record_hparam_flags() as task_flags:
```

Comment:
This seems like an unnecessary holdover from the federated trainer in optimization/.